compose: Redesign the "go to conversation" button.

N-Shar-ma opened the pull request:
Now the button uses an eye icon, is placed at the start of the topic/dm recipient input, and has a bluish purple look. It slides in and out of view as needed.

Fixes: #28697.

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/68962290/b1b96111-ed48-4a88-91c7-2916666533f8)
![image](https://github.com/zulip/zulip/assets/68962290/843b3cf6-6ace-454a-a7a1-89c477d40d87)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

N-Shar-ma commented on web/styles/input_pill.css:
```diff
@@ -120,7 +120,7 @@
 }
 
 #compose-direct-recipient.pill-container {
-    padding: 0 2px;
```
Removing this padding should not be a problem since pills already have margin to space themselves

alya commented:
Playing with the PR, this looks like a nice improvement over the current design!

I don't know how hard it would be to fully implement the design.

alya commented:
@karlstolley Perhaps you could take a look at this next?

karlstolley commented on web/styles/compose.css:
```diff
@@ -188,19 +188,49 @@
-            align-self: center;
+            display: flex;
+            align-items: center;
+            background: hsl(227deg 78% 59% / 12%);
```
It'll probably help the redesign along if you make these color values into CSS variables.

N-Shar-ma replied:
Added variables for all colors in both dark and light theme for the new button!

karlstolley commented on web/styles/compose.css:
```diff
@@ -188,19 +188,49 @@
+            /* Positioned absolutely so it takes up full height but
+            stays inside its flex parent when its other children (pills)
+            wrap onto another row */
+            position: absolute;
```
I kind of don't love this use of relative/absolute positioning. I don't know whether it's worth trying to get it into this PR, but it seems to me like the the container ought to be defined as a grid, with one column for the icon and the other for the pills (which might need their own flex container).

karlstolley replied:
And also, a gridded presentation like this is probably going to be useful for pulling off the overlap-style presentation in the demo.

N-Shar-ma replied:
I've made a prep commit giving the container a grid layout! 

Also, I realised that the transparency of the element overlapping the new button was causing the button to show through, so I changed those transparent colors to solid ones, but they do have the same look as before. @karlstolley Please let me know if that makes sense. Have elaborated a little more in comments in the css.

@alya Now that the slight overlapping works, and things look near identical to the demo, you might want to give this PR another test, or see the updated screenshots in the PR description.

timabbott commented:
Including this in today's chat.zulip.org deployment for further feedback.

N-Shar-ma commented:
Have addressed the feedback from [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/Redesign.20go.20to.20conversation.20button.20.2328915/near/1742249) (added debouncing with delay of 1 second, centered the eye icon, and added more padding for the adjacent topic input)

N-Shar-ma commented:
When focus is in the topic input and the button needs to be shown, it only slides in when focus leaves the topic input, reducing distraction

timabbott commented:
OK updating this deployment on chat.zulip.org.

N-Shar-ma commented:
@zulipbot remove "deployed on chat.zulip.org"

timabbott commented:
Redeploying on chat.zulip.org.

timabbott commented:
Rebased this and am updating the chat.zulip.org deployment.

timabbott commented:
I think the puppeteer test failure might be somehow related to this code, since it's failing on the chat.zulip.org deployment too?

BTW is the first commit a visual noop? Wondering if we can merge that to simplify logistics.

N-Shar-ma commented:
@timabbott Yes, the puppeteer failure was indeed related to this change, have investigated and fixed it now.

And yes, the 1st commit should not cause any visual changes. Though at this point I think the 2nd commit may be good to go too?

N-Shar-ma commented on web/e2e-tests/drafts.test.ts:
```diff
@@ -298,7 +298,7 @@ async function drafts_test(page: Page): Promise<void> {
     // Narrow to the conversation so that the compose box will restore it,
     // then close and try restoring it by opening the composebox again.
-    await page.click("#compose .narrow_to_compose_recipients");
+    await page.click("#compose-direct-recipient .narrow_to_compose_recipients");
```
This more specific selector is needed since we have 2 `.narrow_to_compose_recipients` buttons now, one for dm recipients and one for channel topics, and we need to ensure we click the right one

timabbott commented:
Merged 1ba3cda22950fac965111c9099d9b22ceee7f8cf to main. I'm still not sure we've settled the design for the second commit, so leaving this as an experiment. (I think it implements what we designed).

timabbott commented on web/styles/compose.css:
```diff
@@ -742,63 +768,45 @@ textarea.new_message_textarea {
+        it overlaps, hence we use inherit instead */
+        background: inherit;
+        /* We make the borders of both these elements look like 1
+        single element's border */
```
Generally, comments should be indented, so they don't look like bits of normal code.

timabbott commented on web/styles/compose.css:
```diff
@@ -185,27 +185,44 @@
+            color: var(--color-narrow-to-compose-recipients-color);
+            border: 1px solid var(--color-narrow-to-compose-recipients-border);
+            font-size: 20px;
+            /* extra 3px padding, and -ive 4px margin on the right so that the curved
```
What does `and -ive 4px margin`  mean?

timabbott commented on web/src/compose_recipient.ts:
```diff
@@ -34,6 +34,10 @@ type DirectMessagesOption = {
 
+// This state variable tracks whether the narrow_to_compose_recipients
+// button should be shown when beside the topic input, on blur.
+let show_narrow_to_compose_recipients = true;
```
Can you remind me why we need to track this at all?

timabbott commented:
Posted a few comments, none of which are likely to change the implementation.

N-Shar-ma commented:
Closed in favour of #30027
